### ISSUE TITLE
compareTo and compareToIgnoreCase from String generates ArrayOutOfBoundException

### ISSUE THREAD

**Issue by cedricviaccoz** [scala-native/scala-native#475]
Using in any way the method compareTo from String will result in an ArrayOutOfBoundException. 
This is a common mistake made while making ports of Java code to Scala. 
On [String.java](https://github.com/apache/harmony/blob/java6/classlib/modules/luni/src/main/java/java/lang/String.java) from harmony on line 729 we have : 
```java
if ((c1 = value[o1++]) == (c2 = target[o2++])) {
```
which was ported as 
```scala 
o1 += 1
o2 += 1 
val c1: Char    = compareValue(value(o1))
val c2: Char    = compareValue(string.value(o2))
```
However in java, `idx++` is an expression which returns the value of `idx` first, and then increments it. So in an array access, `array[idx++]` means we access the array at idx position, and then we increments idx.
Thus the good way to port the expression `array[idx++]` in scala would be 
```scala
array(idx)
idx += 1
```
And in the case of String.scala, the good way to write this part of the port would thus be 
```scala
val c1: Char    = compareValue(value(o1))
val c2: Char    = compareValue(string.value(o2))
o1 += 1
o2 += 1 
```
Explaining thus ArrayOutOfBoundExceptions, since we incremented before accessing.
Regarding only String.scala, this issue is present in methods compareTo, compareToIgnoreCase, equalsIgnoreCase , getBytes, regionMatches, and replace.

**Comment by cedricviaccoz** [comment]
PR  #476 was opened to correct those errors

**Comment by densh** [comment]
@cedricviaccoz Thanks for the report and investigation!

**Comment by densh** [comment]
Fixed in #476
